### PR TITLE
Review

### PR DESCRIPTION
**Pull request by arpita-biswas**

### CODE REVIEW

File: Tree.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+		//System.out.println("Root: "+this.root.data);
+		createTree(this.root, 0, data);				
+	}
+	
+	private void createTree(Node<Integer> root, int n, int data[]){
+		int leftN = 2*n + 1;
+		int rightN = leftN + 1;
+		if(leftN >= data.length){
+			//System.out.println("Left of "+data[n]+" is not there and exceeds length");
+			root.left = null;
```
Comment (Thomas-de-Rivaz) [inline Tree.java]: I don't think you need to set the value of left and right, they will be null by default.

File: Tree.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+			return;
+		}
+		else{
+			if(data[leftN] != NullNode){		
+				Node<Integer> left = new Node<Integer>(data[leftN]);
+				//System.out.println("Left of "+data[n]+" is "+data[leftN]);
+				root.left = left;
+				createTree(left, leftN, data);
```
Comment (Thomas-de-Rivaz) [inline Tree.java:45]: How deep a tree do you think you could create with recursion? If I passed in an input of 0,1,MIN_INT,2,MIN_INT,....,1000,MIN_INT,MIN_INT. Would that work? If not why?

File: Tree.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	
+	private StringBuilder print(Node<Integer> root){
+		StringBuilder str = new StringBuilder();
+		if(root == null){
+			return str;
+		}
+		
+		str.append("Node: "+root.data+" ");
+		if(root.left != null){
+			str.append("Left: "+root.left.data+" ");
```
Comment (Thomas-de-Rivaz) [inline Tree.java]: If Node itself is given a toString method, this code will get a lot cleaner. Tree shouldn't care how to print an individual node.

File: Q1.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+			str.append("Empty tree");
+			return str.toString();
+		}
+		if(!allAncestorHelper(t.root, num, str)){
+			str.append(num+" not found");
+		}
+		return str.toString();
+	}
+	
+	static boolean allAncestorHelper(Node<Integer> root, int num, StringBuilder str){
```
Comment (Thomas-de-Rivaz) [inline Q1.java]: allAncestorHelper is a complex method. Adding a comment explaining what will be appended to str would be helpful

File: Q1.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+			str.append(root.data+", ");
+			return true;
+		}
+		return false;
+	}
+	
+	public static void main(String[] args) {
+		
+		//Tests
+		testNullTree(); //When (t == null)
```
Comment (Thomas-de-Rivaz) [inline Q1.java:64]: Good test choices

File: Q2.java [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+		return false;
+	}
+	
+	public static void main(String[] args) {
+		
+		//Tests
+		testNullTree(); //When (t == null)
+		testEmptyTree(); //When (t.root == null)
+		testElementsNotPresent(); //When p, q or both not present in the tree
+		testSamePQ(); //When p and q
```
Comment (Thomas-de-Rivaz) [inline Q2.java:137]: Good Tests

File: Q2.java [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+			str.append("Lowest Common Ancestor of "+p+" and "+q+" is "+res.potentialAncestor.data);
+			return str.toString();
+		}
+		else{
+			str.append("Either "+p+" or "+q+" is not present");
+			return str.toString();
+		}
+	}
+	
+	static Result commonAncestorHelper(Node<Integer> root, int p, int q){
```
Comment (Thomas-de-Rivaz) [inline Q2.java]: Add a comment to commonAncestorHelper detailing what the returned Result means.

File: Q2.java [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+				
+		//if root is null, then return null node as potentialAncestor with isAncestor = false
+		if(root == null){
+			return new Result(null, false);
+		}
+		//System.out.println("Node: "+root.data);
```
Comment (Thomas-de-Rivaz) [inline Q2.java:70]: Remember to delete all of the System.outs you used for testing.

File: Q2.java [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+		
+		//Check if root contains p
+		if(root.data == p){//if root contains p, then search for q in the subtree
```
Comment (Thomas-de-Rivaz) [inline Q2.java]: There is quite a bit of duplication with p & q swapped.

**Review by Thomas-de-Rivaz — APPROVED** [review]
What do you think could be modified in assignment 2 to make it work for finding the lowest common ancestor of more than 2 nodes?

File: Tree.java (hunk 2 above)
Comment (arpita-biswas) [reply Tree.java:45]: In worst case (right skewed tree), for populating n elements, an array of 2^n -1 elements is required (which is not at all memory-efficient). In this case, I feel it is better to use adjacency list as input for populating the tree. Shall I re-write the code such that it also takes an adjacency list as input? Are you pointing to any other reason for it to not work?

File: Q1.java (hunk 4 above)
Comment (arpita-biswas) [reply Q1.java]: Done

File: Q1.java (hunk 5 above)
Comment (arpita-biswas) [reply Q1.java:64]: Thank you

File: Q2.java (hunk 7 above)
Comment (arpita-biswas) [reply Q2.java]: Done

File: Q2.java (hunk 8 above)
Comment (arpita-biswas) [reply Q2.java:70]: Done

File: Q2.java (hunk 9 above)
Comment (arpita-biswas) [reply Q2.java]: I have created three cases now:
If root = p = q: 
Else if root = p or root = q: 
Else:

File: Q2.java (hunk 6 above)
Comment (arpita-biswas) [reply Q2.java:137]: Thank you

**Pull request merged**